### PR TITLE
[config] Deprecate component.UnmarshalConfig

### LINKED ISSUE #7101: Investigate alternative for `Unmarshaler` and `Marshaler`
The `Unmarshaler` because of the `Unmarshal` definition, has an issue when recursively unmarshal other structs that implement the `Unmarshaler`, see https://github.com/open-telemetry/opentelemetry-collector/pull/6029 where the "given" destination to Conf.Unmarshal cannot be checked against the Unmarshaler implementation to avoid a deadlock. The Marshaler seem to not have (I did not carefully check) the same problem, but this needs to be confirmed.

Some possible solutions:
* See yaml.v2 interfaces, they seem to work fine. We may want to switch to that.
* Ignore this issue, document properly in the Conf.Unmarshal the caveat.

### LINKED ISSUE #7102: Remove component.UnmarshalConfig
There are mostly historical reasons why this exists, but we should clean [this](https://github.com/open-telemetry/opentelemetry-collector/blob/main/component/config.go#L41). Here are few options:

* Provide a `confmap.Unmarshal` equivalent with `component.UnmarshalConfig`. Few caveats here:
  * The issue with Conf.Umarshal https://github.com/open-telemetry/opentelemetry-collector/issues/7101 which kind of require to have a separate func outside of Conf to avoid the circular call.
  * The default "Options" for Unmarshal are different, maybe we can fix this by making the `ErrorUnused` enable by default.
* Fix https://github.com/open-telemetry/opentelemetry-collector/issues/7101 and default enable `ErrorUnused`. This will become super clean, the Conf.Unmarshal can be used, offers options for customizations, etc.

### PR DESCRIPTION
**Pull request by atoulme**
**Description:**
This PR removes the top level if/else in `component.UnmarshalConfig`, handling recursive state in the confmap.Conf object instead.
This PR deprecates `component.UnmarshalConfig` in favor of calling directly `Unmarshal` on the confmap.Conf object.

**Link to tracking Issue:**
Fixes #7102
Fixes #7101

### CODE REVIEW

**Comment by atoulme** [comment]
Contrib tests fix: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31727

File: cmd/mdatagen/loader.go [hunk 1]
```diff
@@ -124,11 +124,7 @@ func (m *metric) Unmarshal(parser *confmap.Conf) error {
 	if !parser.IsSet("enabled") {
 		return errors.New("missing required field: `enabled`")
 	}
-	err := parser.Unmarshal(m)
```
Comment (mx-psi) [inline cmd/mdatagen/loader.go:127]: I think this one I already merged on a different PR, right? Can you rebase your branch?

File: cmd/mdatagen/metricdata.go
Comment (mx-psi) [inline cmd/mdatagen/metricdata.go:1]: Can we do all changes in this file in a different PR? AFAICT they can be done independently

File: component/config_test.go [hunk 2]
```diff
@@ -419,28 +417,3 @@ func TestNewType(t *testing.T) {
...
-	embeddedUnmarshallingConfig
-}
-
-type embeddedUnmarshallingConfig struct {
-}
-
-func (euc *embeddedUnmarshallingConfig) Unmarshal(_ *confmap.Conf) error {
-	return nil // do nothing.
-}
-func TestStructWithEmbeddedUnmarshaling(t *testing.T) {
```
Comment (mx-psi) [inline component/config_test.go:435]: Why are we removing this test? Is it being move somewhere else?

File: confmap/confmap.go [hunk 3]
```diff
@@ -143,31 +147,40 @@ func (l *Conf) ToStringMap() map[string]any {
...
+		mapstructure.StringToSliceHookFunc(","),
+		mapKeyStringToMapKeyTextUnmarshalerHookFunc(),
+		mapstructure.StringToTimeDurationHookFunc(),
+		mapstructure.TextUnmarshallerHookFunc(),
+		unmarshalerHookFunc(result, topLevelUnmarshaling),
+		// after the main unmarshaler hook is called,
+		// we unmarshal the embedded structs if present to merge with the result:
+		unmarshalerEmbeddedStructsHookFunc(),
+		zeroSliceHookFunc(),
+	}
```
Comment (mx-psi) [inline confmap/confmap.go:162]: This is also something we can do on a separate PR!

File: confmap/confmap.go [hunk 4]
```diff
@@ -37,7 +40,8 @@ func NewFromStringMap(data map[string]any) *Conf {
 // Conf represents the raw configuration map for the OpenTelemetry Collector.
 // The confmap.Conf can be unmarshalled into the Collector's config using the "service" package.
 type Conf struct {
-	k *koanf.Koanf
+	k    *koanf.Koanf
+	self any
```
Comment (mx-psi) [inline confmap/confmap.go:44]: Before I think we could call the same `Conf` to unmarshal it on two different structs, but now it can only be used once, right? Is there any way we can restore that property? If not, is there a way where we can error out if we have already unmarshaled?

File: component/config_test.go (hunk 2 above)
Comment (atoulme) [reply component/config_test.go:435]: It's moved under confmap.

File: confmap/confmap.go (hunk 3 above)
Comment (atoulme) [reply confmap/confmap.go:162]: right, I'll revert that change. It's not helping the diff.

File: cmd/mdatagen/loader.go (hunk 1 above)
Comment (atoulme) [reply cmd/mdatagen/loader.go]: Done, thanks

File: confmap/confmap.go (hunk 4 above)
Comment (atoulme) [reply confmap/confmap.go:44]: This doesn't break your ability to use the same `Conf` to unmarshal on two different structs. It stops recursive calls as part of unmarshaling from calling `Unmarshal` on the same object twice.

File: confmap/confmap.go [hunk 5]
```diff
@@ -167,7 +171,13 @@ func decodeConfig(m *Conf, result any, errorUnused bool) error {
 	if err != nil {
 		return err
 	}
-	return decoder.Decode(m.ToStringMap())
+	if err = decoder.Decode(m.ToStringMap()); err != nil {
```
Comment (atoulme) [inline confmap/confmap.go:174]: Splitting this change into #9763 as well.

File: component/config_test.go (hunk 2 above)
Comment (atoulme) [reply component/config_test.go:435]: brought it back with a few changes so it passes!

**Comment by mx-psi** [comment]
Can you add 'Fixes #7101' to the PR description?

File: confmap/confmap.go [hunk 6]
```diff
@@ -310,33 +315,47 @@ func unmarshalerEmbeddedStructsHookFunc() mapstructure.DecodeHookFuncValue {
...
-		if _, ok = from.Interface().(map[string]any); !ok {
-			return from.Interface(), nil
+		unmarshalMethod := reflect.ValueOf(toPtr).MethodByName("Unmarshal")
+		// check that the Unmarshal method is not defined on one of the squash embedded structs defined.
+		// This allows us to have composition where the Unmarshaler may embed structs implementing Unmarshaler.
+		if to.Type().Kind() == reflect.Struct {
+			for i := 0; i < to.Type().NumField(); i++ {
+				// check embedded structs:
+				f := to.Type().Field(i)
+				if f.IsExported() && f.Anonymous && slices.Contains(strings.Split(f.Tag.Get("mapstructure"), ","), "squash") {
```
Comment (dmitryax) [inline confmap/confmap.go:341]: Why do we need to check for the `mapstructure` tag?

Comment (atoulme) [reply confmap/confmap.go:341]: This is explained in the comment above: if the embedded struct is using the `,squash` value for the `mapstructure` key, then this struct Unmarshal method should not be used as the main struct unmarshaling.

File: .chloggen/embedded_struct_unmarshaler.yaml [hunk 7]
```diff
@@ -0,0 +1,25 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: breaking
```
Comment (andrzej-stencel) [inline .chloggen/embedded_struct_unmarshaler.yaml:4]: Isn't this a `deprecation`?

File: component/config.go [hunk 8]
```diff
@@ -27,13 +27,8 @@ type Config any
 var configValidatorType = reflect.TypeOf((*ConfigValidator)(nil)).Elem()
 
 // UnmarshalConfig helper function to UnmarshalConfig a Config.
-// It checks if the config implements confmap.Unmarshaler and uses that if available,
-// otherwise uses Map.UnmarshalExact, erroring if a field is nonexistent.
+// Deprecated: [v0.101.0] Use conf.Unmarshal(&intoCfg)
 func UnmarshalConfig(conf *confmap.Conf, intoCfg Config) error {
```
Comment (andrzej-stencel) [inline component/config.go:31]: Do we want to change the way this function works with the deprecation? Is there a reason we cannot leave the function unchanged until it's completely removed?

File: .chloggen/embedded_struct_unmarshaler.yaml (hunk 7 above)
Comment (atoulme) [reply .chloggen/embedded_struct_unmarshaler.yaml:4]:
```suggestion
change_type: deprecation
```

File: component/config.go (hunk 8 above)
Comment (atoulme) [reply component/config.go:31]: It does, it removes the old way to call into unmarshaling by calling the Unmarshaler Unmarshal function.

**Review by mx-psi — APPROVED** [review]

File: component/config.go (hunk 8 above)
Comment (mx-psi) [reply component/config.go:31]: AIUI this is a no-op change

File: .chloggen/embedded_struct_unmarshaler.yaml [hunk 9]
```diff
@@ -0,0 +1,25 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: deprecation
+
+# The name of the component, or a single word describing the area of concern, (e.g. otlpreceiver)
+component: component
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Deprecate `component.UnmarshalConfig`, use `conf.Unmarshal(&intoCfg)` instead.
```
Comment (evan-bradley) [inline .chloggen/embedded_struct_unmarshaler.yaml:10]:
```suggestion
note: Deprecate `component.UnmarshalConfig`, use `(*confmap.Conf).Unmarshal(&intoCfg)` instead.
```
Up to you if this is more readable, I just want to make it clear we're talking about a method on a `confmap.Conf` object and not a `conf` package, since we use the word `config` in so many different contexts in our APIs.

**Review by evan-bradley — APPROVED** [review]
